src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -245,113 +193,87 @@ bool RandomFourierGaussPreproc::init_randomcoefficients() {
 			}
 
 			// =  x1/std::sqrt(val)* std::sqrt(-2*std::log(val));
-			randomcoeff_multiplicative[i * cur_dim_input_space + k] =
-			    x1 * std::sqrt(-2 * std::log(s) / s) / kernelwidth;
+			m_randomweights[i * dim_input_space + k] =
+			    x1 * std::sqrt(-2 * std::log(s) / s) / m_kernelwidth;
 		}
 	}
 
-	io::info("finished: initializing randomcoefficients ");
-
-	return true;
-}
-
-void RandomFourierGaussPreproc::get_randomcoefficients(
-		float64_t ** randomcoeff_additive2,
-		float64_t ** randomcoeff_multiplicative2, int32_t *dim_feature_space2,
-		int32_t *dim_input_space2, float64_t* kernelwidth2) const {
-
-	ASSERT(randomcoeff_additive2)
-	ASSERT(randomcoeff_multiplicative2)
-
-	if (!test_rfinited()) {
-		*dim_feature_space2 = 0;
-		*dim_input_space2 = 0;
-		*kernelwidth2=1;
-		*randomcoeff_additive2 = NULL;
-		*randomcoeff_multiplicative2 = NULL;
-		return;
-	}
-
-	*dim_feature_space2 = cur_dim_feature_space;
-	*dim_input_space2 = cur_dim_input_space;
-	*kernelwidth2=cur_kernelwidth;
-
-	*randomcoeff_additive2 = SG_MALLOC(float64_t, cur_dim_feature_space);
-	*randomcoeff_multiplicative2 = SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
-
-	std::copy(randomcoeff_additive, randomcoeff_additive+cur_dim_feature_space,
-			*randomcoeff_additive2);
-	std::copy(randomcoeff_multiplicative, randomcoeff_multiplicative+cur_dim_feature_space*cur_dim_input_space,
-			*randomcoeff_multiplicative2);
+	io::info("finished: initializing randcoefficients ");
 
+	is_updated = true;
 
+	return true;
 }
 
-void RandomFourierGaussPreproc::set_randomcoefficients(
-		float64_t *randomcoeff_additive2,
-		float64_t * randomcoeff_multiplicative2,
-		const int32_t dim_feature_space2, const int32_t dim_input_space2, const float64_t kernelwidth2) {
-	dim_feature_space = dim_feature_space2;
-	dim_input_space = dim_input_space2;
-	kernelwidth=kernelwidth2;
-
-	SG_FREE(randomcoeff_multiplicative);
-	randomcoeff_multiplicative=NULL;
-	SG_FREE(randomcoeff_additive);
-	randomcoeff_additive=NULL;
+void RandomFourierGaussPreproc::extract_coefficients(
+    const RandomFourierGaussPreproc& other)
+{
+	m_dim_feature_space = other.dim_feature_space;
+	m_kernelwidth = other.m_kernelwidth;
 
-	cur_dim_feature_space=dim_feature_space;
-	cur_dim_input_space = dim_input_space;
-	cur_kernelwidth=kernelwidth;
+	SG_FREE(m_randomweights);
+	m_randomweights = NULL;
+	SG_FREE(m_randomoffset);
+	m_randomoffset = NULL;
 
-	if( (dim_feature_space>0) && (dim_input_space>0) )
+	if (m_dim_feature_space > 0)
 	{
-	randomcoeff_additive=SG_MALLOC(float64_t, cur_dim_feature_space);
-	randomcoeff_multiplicative=SG_MALLOC(float64_t, cur_dim_feature_space*cur_dim_input_space);
 
-	std::copy(randomcoeff_additive2, randomcoeff_additive2
-			+ dim_feature_space, randomcoeff_additive);
-	std::copy(randomcoeff_multiplicative2, randomcoeff_multiplicative2
-			+ cur_dim_feature_space*cur_dim_input_space, randomcoeff_multiplicative);
+		m_randomweights =
```

Comment:
see above comments.
You can avoid such things if you read a bit of shogun source code. There isnt a single place where this is done

Reply:
My bad, I only pushed the code so that you can help with the part I highlighted. I'll be careful in the next push to avoid wasting our time.